tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,511 @@
+            {
+                var dynMethod = new DynamicMethod(
+                    "OnSampleDyn",
+                    activitySamplingResultType,
+                    new[] { activityCreationOptionsType.MakeGenericType(activityContextType).MakeByRefType() },
+                    typeof(ActivityListener).Module,
+                    true);
+
+                var il = dynMethod.GetILGenerator();
+                il.EmitCall(OpCodes.Call, onSampleMethodInfo, null);
+                il.Emit(OpCodes.Ret);
+
+                return dynMethod.CreateDelegate(sampleActivityType.MakeGenericType(activityContextType));
+            }
+
+            public static Delegate CreateOnSampleUsingParentIdDelegate(Type activitySamplingResultType, Type activityCreationOptionsType, Type sampleActivityType, MethodInfo onSampleUsingParentIdMethodInfo)
+            {
+                var dynMethod = new DynamicMethod(
+                    "OnSampleUsingParentIdDyn",
+                    activitySamplingResultType,
+                    new[] { activityCreationOptionsType.MakeGenericType(typeof(string)).MakeByRefType() },
+                    typeof(ActivityListener).Module,
+                    true);
+
+                var il = dynMethod.GetILGenerator();
+                il.EmitCall(OpCodes.Call, onSampleUsingParentIdMethodInfo, null);
+                il.Emit(OpCodes.Ret);
+
+                return dynMethod.CreateDelegate(sampleActivityType.MakeGenericType(typeof(string)));
+            }
+
+            public static Delegate CreateOnShouldListenToDelegate(Type activitySourceType, MethodInfo onShouldListenToMethodInfo)
+            {
+                var dynMethod = new DynamicMethod(
+                    "OnShouldListenToDyn",
+                    typeof(bool),
+                    new[] { activitySourceType },
+                    typeof(ActivityListener).Module,
+                    true);
+
+                var proxyResult = DuckType.GetOrCreateProxyType(typeof(IActivitySource), activitySourceType);
+                var method = onShouldListenToMethodInfo.MakeGenericMethod(proxyResult.ProxyType);
+                var proxyTypeCtor = proxyResult.ProxyType.GetConstructors()[0];
+
+                var il = dynMethod.GetILGenerator();
+                il.Emit(OpCodes.Ldarg_0);
+                il.Emit(OpCodes.Newobj, proxyTypeCtor);
+                il.EmitCall(OpCodes.Call, method, null);
+                il.Emit(OpCodes.Ret);
+
+                return dynMethod.CreateDelegate(typeof(Func<,>).MakeGenericType(activitySourceType, typeof(bool)));
+            }
+        }
+
+        internal class DiagnosticSourceEventListener : IObserver<KeyValuePair<string, object>>
+        {
+            private readonly string _sourceName;
+
+            private DiagnosticSourceEventListener(string sourceName)
+            {
+                _sourceName = sourceName;
+            }
+
+            internal static void OnSetListener(object value)
```

Comment:
Can we refactor the parameter name to `diagnosticListener` or `listener`? It also might be worth adding some comments or link to the [DiagnosticSourceUsersGuide](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md#consuming-data-with-diagnosticlistener) to help guide new reviewers through this machinery. I have to admit it took me a long time to grok my head around all the reflection going on in this initializiation 😅

Reply:
I have two other suggestions, let me know what you think of them:

1. Maybe this method could also be renamed to something like `DynamicDiagnosticObserver_OnNext`, to make it a little more clear that this is invoked from the dynamic `IObserver<DiagnosticListener>` object that's created
2. Maybe this method could be moved up into the owning type. That way we reduce a little of the field leakage that Andrew referenced and we don't confuse that observer with the DiagnosticListener's observer

Reply:
I've created a new `DiagnosticObserverListener` class as part of the refactor.